examples/HTMLBuilder.swift

```diff
@@ -0,0 +1,330 @@
+    into stream: HTMLOutputStream,
+    attributes: [Attribute]
+  ) {
+
+    // Write the opening of the tag, e.g. `<div`
+    stream.writeIndent()
+    stream.write("<")
+    stream.write(tag)
+
+    // Write each attribute, e.g. `class="my-class"`.
+    for attr in attributes {
+      stream.write(" ")
+      stream.write(attr.key)
+
+      // Some attributes do not have values, and instead use the presence of the
+      // for its effects, such as `<input disabled>`.
+      if let value = attr.value {
+        stream.write("=")
+        stream.writeDoubleQuoted(value)
+      }
+    }
+
+    // If this node has no children, end the tag without introducing a second
+    // closing tag, e.g. `<br />`.
+    guard let child = child else {
+      stream.write("/>\n")
+      return
+    }
+
+    // Otherwise, add a newline and render the child indented.
+    stream.write(">\n")
+    stream.withIndent {
+      // Explicitly do not pass any attributes to children.
+      child.renderAsHTML(into: stream, attributes: [])
+    }
+
+    // Write the closing tag on its own line.
+    stream.writeIndent()
+    stream.write("</")
+    stream.write(tag)
+    stream.write(">\n")
+  }
+}
+
+// MARK: - Concatenative HTML Attributes
+
+public struct Attribute {
+  let key: String
+  let value: String?
+}
+
+/// Wraps an HTML object with a given attribute. these attributes are collected
+/// while walking the node hierarchy and finally printed when printing an
+/// `HTMLNode`.
+struct AttributedNode: HTML {
+  let attribute: Attribute
+  let child: HTML
+
+  func renderAsHTML(into stream: HTMLOutputStream, attributes: [Attribute]) {
+    var fullAttrs = attributes
+    fullAttrs.append(attribute)
+    child.renderAsHTML(into: stream, attributes: fullAttrs)
+  }
+}
+
+/// Iterates over a collection of data and applies a transformation to each
+/// piece of data that creates an HTML node.
+public func forEach<Coll: Collection>(
+  _ data: Coll,
+  @HTMLBuilder content: @escaping (Coll.Element) -> HTML
+) -> HTML {
+  return MultiNode(children: data.map(content))
+}
+
+// MARK: - HTMLBuilder
+
+/// A function builder that provides transformations for control flow concepts
+/// into HTML components.
+@_functionBuilder
+public struct HTMLBuilder {
+  /// If there are no children in an HTMLBuilder closure, then return an empty
+  /// MultiNode.
+  public static func buildBlock() -> HTML {
+    return MultiNode(children: [])
+  }
+
+  /// If there is one child, return it directly.
+  public static func buildBlock(_ content: HTML) -> HTML {
+    return content
+  }
+
+  /// If there are multiple children, return them all as a MultiNode.
+  public static func buildBlock(_ content: HTML...) -> HTML {
+    return MultiNode(children: content)
+  }
+
+  /// If the provided child is `nil`, build an empty MultiNode. Otherwise,
+  /// return the wrapped value.
+  public static func buildIf(_ content: HTML?) -> HTML {
+    if let content = content { return content }
```

Comment:
there is probably a typo somewhere here, this will return nil, but the return type is not an optional